crates/dynamodb-streams/src/stream.rs

```diff
@@ -163,20 +162,96 @@ impl DynamodbStreamProducer {
     }
 }
 
-fn combine_shard_batches(batches: Vec<ShardPollResult>) -> DynamoDBStreamBatch {
+fn combine_shard_batches(
+    poll_results: &[ShardPollResult],
+    idle_timeout: Option<Duration>,
+) -> DynamoDBStreamBatch {
+    let now = SystemTime::now();
+
+    // Collect records, checkpoints and watermarks
     let mut records = Vec::new();
+    let mut shard_watermarks = Vec::new();
     let mut shards_checkpoints = HashMap::new();
 
-    for batch in batches {
-        records.extend(batch.records);
-        shards_checkpoints.insert(batch.shard_id, batch.checkpoint);
+    for shard_result in poll_results {
+        // Collect records
+        if let PollOutcome::Records {
+            records: shard_records,
+        } = &shard_result.outcome
+        {
+            records.extend(shard_records.clone());
+        }
+
+        // Collect checkpoints
+        shards_checkpoints.insert(
+            shard_result.shard_id.clone(),
+            shard_result.last_checkpoint.clone(),
+        );
+
+        // Check eligibility for watermark
+        let is_eligible = match shard_result.outcome {
+            // Shards that produced records and those that failed are always eligible
+            PollOutcome::Records { .. } | PollOutcome::Failed => true,
+
+            // Shards that produced no records are NOT eligible if they have been idle for longer than the idle_timeout
+            PollOutcome::Empty => {
```

Comment:
Nit: review if PollOutcome::Empty actually means this is shard end so we don't need 'idle timeout' logic

Reply:
Will look into it, thanks!